### ISSUE TITLE
DNS rules don't work anymore on kubernetes 1.18 with cilium 1.8

### ISSUE THREAD

**Issue by gandalfmagic** [cilium/cilium#13308]
## Bug report

**General Information**

- Cilium version `1.8.3 54cf3810d`
- Kernel version `Linux node1 5.7.0-0.bpo.2-amd64 #1 SMP Debian 5.7.10-1~bpo10+1 (2020-07-30) x86_64 GNU/Linux`
- Orchestration system `Kubernetes v1.18.9`

**Description of the problem**

After upgrading a Kubernetes cluster from v1.17.7 to v1.18.9 and Cilium from 1.7.5 to v1.8.3 the policies containing a DNS rule don't work anymore.

This is a test manifest I use to reproduce the problem:

```yaml
---
apiVersion: v1
kind: Namespace
metadata:
  name: test
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: test
  name: test
  namespace: test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test
  template:
    metadata:
      labels:
        app: test
    spec:
      containers:
      - name: test
        image: debian
        command:
        - sleep
        - "20000"
---
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: egress-test
  namespace: test
spec:
  endpointSelector:
    matchLabels:
      "k8s:app": test
  egress:
  - toCIDR:
    - 169.254.25.10/32
    toPorts:
    - ports:
      - port: "53"
        protocol: UDP
      rules:
        dns:
        - matchPattern: "*"
```

If I try to ping an outside DNS name, I receive from the container, I get the following error:

```bash
# ping www.google.com
ping: www.google.com: Temporary failure in name resolution
```

And the `fqdn cache` is empty.

The same manifests, on a cluster with k8s v1.17.7 and cilium 1.7.5 works just fine:

```bash
# ping www.google.com                    
PING www.google.com (216.58.198.36) 56(84) bytes of data.
```

And the `fqdn cache` is populated:

```bash
$ k8s-cilium-exec.sh cilium fqdn cache list
Endpoint   Source       FQDN                                                   TTL    ExpirationTime             IPs                        
3018       lookup       www.google.com.                                        3600   2020-09-28T11:12:20.121Z   2a00:1450:4002:802::2004   
3018       lookup       www.google.com.                                        3600   2020-09-28T11:12:20.121Z   216.58.198.36              
```

If I try to monitor the policy of the non working pod, here is the results:

```bash
$ k8s-cilium-exec.sh cilium monitor --related-to 2902
Policy verdict log: flow 0x6830825c local EP ID 2902, remote ID 16777217, dst port 53, proto 17, ingress false, action redirect, match L3-L4, 10.233.66.224:50827 -> 169.254.25.10:53 udp
-> proxy flow 0x6830825c identity 2950->0 state new ifindex 0 orig-ip 0.0.0.0: 10.233.66.224:50827 -> 169.254.25.10:53 udp
-> proxy flow 0x6830825c identity 2950->0 state established ifindex 0 orig-ip 0.0.0.0: 10.233.66.224:50827 -> 169.254.25.10:53 udp
-> endpoint 2902 flow 0x0 identity 16777217->2950 state reply ifindex lxcb73878e9d0fa orig-ip 169.254.25.10: 169.254.25.10:53 -> 10.233.66.224:50827 udp
level=info msg="Initializing dissection cache..." subsys=monitor
-> endpoint 2902 flow 0x0 identity 16777217->2950 state reply ifindex lxcb73878e9d0fa orig-ip 169.254.25.10: 169.254.25.10:53 -> 10.233.66.224:50827 udp
-> proxy flow 0x6830825c identity 2950->0 state established ifindex 0 orig-ip 0.0.0.0: 10.233.66.224:50827 -> 169.254.25.10:53 udp
-> proxy flow 0x6830825c identity 2950->0 state established ifindex 0 orig-ip 0.0.0.0: 10.233.66.224:50827 -> 169.254.25.10:53 udp
-> Request dns from 2902 ([k8s:io.kubernetes.pod.namespace=test k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.cilium.k8s.policy.cluster=default k8s:app=test]) to 0 ([reserved:world]), identity 2950->2, verdict Denied DNS Query: www.googlc.ome.test.svc.cluster.local. A
-> endpoint 2902 flow 0x0 identity 16777217->2950 state reply ifindex lxcb73878e9d0fa orig-ip 169.254.25.10: 169.254.25.10:53 -> 10.233.66.224:50827 udp
-> Request dns from 2902 ([k8s:io.cilium.k8s.policy.cluster=default k8s:app=test k8s:io.kubernetes.pod.namespace=test k8s:io.cilium.k8s.policy.serviceaccount=default]) to 0 ([reserved:world]), identity 2950->2, verdict Denied DNS Query: www.googlc.ome.test.svc.cluster.local. AAAA
-> endpoint 2902 flow 0x0 identity 16777217->2950 state reply ifindex lxcb73878e9d0fa orig-ip 169.254.25.10: 169.254.25.10:53 -> 10.233.66.224:50827 udp
-> Request dns from 2902 ([k8s:io.cilium.k8s.policy.cluster=default k8s:app=test k8s:io.kubernetes.pod.namespace=test k8s:io.cilium.k8s.policy.serviceaccount=default]) to 0 ([reserved:world]), identity 2950->2, verdict Denied DNS Query: www.googlc.ome.test.svc.cluster.local. A
-> Request dns from 2902 ([k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.cilium.k8s.policy.cluster=default k8s:app=test k8s:io.kubernetes.pod.namespace=test]) to 0 ([reserved:world]), identity 2950->2, verdict Denied DNS Query: www.googlc.ome.test.svc.cluster.local. AAAA
Policy verdict log: flow 0x32eed54d local EP ID 2902, remote ID 16777217, dst port 53, proto 17, ingress false, action redirect, match L3-L4, 10.233.66.224:55978 -> 169.254.25.10:53 udp
-> proxy flow 0x32eed54d identity 2950->0 state new ifindex 0 orig-ip 0.0.0.0: 10.233.66.224:55978 -> 169.254.25.10:53 udp
-> proxy flow 0x32eed54d identity 2950->0 state established ifindex 0 orig-ip 0.0.0.0: 10.233.66.224:55978 -> 169.254.25.10:53 udp
-> Request dns from 2902 ([k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.cilium.k8s.policy.cluster=default k8s:app=test k8s:io.kubernetes.pod.namespace=test]) to 0 ([reserved:world]), identity 2950->2, verdict Denied DNS Query: www.googlc.ome. AAAA
-> Request dns from 2902 ([k8s:app=test k8s:io.kubernetes.pod.namespace=test k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.cilium.k8s.policy.cluster=default]) to 0 ([reserved:world]), identity 2950->2, verdict Denied DNS Query: www.googlc.ome. A
-> proxy flow 0x32eed54d identity 2950->0 state established ifindex 0 orig-ip 0.0.0.0: 10.233.66.224:55978 -> 169.254.25.10:53 udp
-> proxy flow 0x32eed54d identity 2950->0 state established ifindex 0 orig-ip 0.0.0.0: 10.233.66.224:55978 -> 169.254.25.10:53 udp
-> endpoint 2902 flow 0x0 identity 16777217->2950 state reply ifindex lxcb73878e9d0fa orig-ip 169.254.25.10: 169.254.25.10:53 -> 10.233.66.224:55978 udp
-> endpoint 2902 flow 0x0 identity 16777217->2950 state reply ifindex lxcb73878e9d0fa orig-ip 169.254.25.10: 169.254.25.10:53 -> 10.233.66.224:55978 udp
-> endpoint 2902 flow 0x0 identity 16777217->2950 state reply ifindex lxcb73878e9d0fa orig-ip 169.254.25.10: 169.254.25.10:53 -> 10.233.66.224:55978 udp
-> endpoint 2902 flow 0x0 identity 16777217->2950 state reply ifindex lxcb73878e9d0fa orig-ip 169.254.25.10: 169.254.25.10:53 -> 10.233.66.224:55978 udp
-> Request dns from 2902 ([k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.cilium.k8s.policy.cluster=default k8s:app=test k8s:io.kubernetes.pod.namespace=test]) to 0 ([reserved:world]), identity 2950->2, verdict Denied DNS Query: www.googlc.ome. AAAA
-> Request dns from 2902 ([k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.cilium.k8s.policy.cluster=default k8s:app=test k8s:io.kubernetes.pod.namespace=test]) to 0 ([reserved:world]), identity 2950->2, verdict Denied DNS Query: www.googlc.ome. A 
```

**How to reproduce the issue**

Use kubespray to deploy a kubernetes cluster on Debian10 with backported kernel .7.0-0.bpo.2-amd64

Deploy the cluster with:
- kube_proxy_mode: iptables
- enable_nodelocaldns: true

Here the relevant configuration files for Kubuspray:
[kubespray-files.zip](https://github.com/cilium/cilium/files/5292073/kubespray-files.zip)

**Comment by joestringer** [comment]
Thanks for the report!

This is interesting:

```
-> proxy flow 0x32eed54d identity 2950->0 state new ifindex 0 orig-ip 0.0.0.0: 10.233.66.224:55978 -> 169.254.25.10:53 udp
```
...
```
-> Request dns from 2902 ([k8s:app=test k8s:io.kubernetes.pod.namespace=test k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.cilium.k8s.policy.cluster=default]) to 0 ([reserved:world]), identity 2950->2, verdict Denied DNS Query: www.googlc.ome. A
```

The succession of these entries implies that the `-> proxy` packet destined to the IP `169.254.25.10` is being redirected towards the proxy, and the latter message from the proxy reports that the traffic does not match a DNS policy. This suggests that either the proxy is not being configured with the policy correctly, or the identity lookup is wrong (I note it reports `identity 2950->2`, yet we have this IP in the policy and hence I would expect the destination identity to be something like `16777217` rather than `2`.

Are you able to provide a sysdump? This makes debugging issues like this much easier as sysdump will gather various details about the configuration in the cluster and any kind of policies, logs with warnings, and Cilium state. One area I'm curious about specifically is the contents of `cilium ip list` as it seems like the DNS proxy lookup is not finding the identity corresponding to the IP.

**Comment by gandalfmagic** [comment]
Here you go.

[cilium-sysdump-20200930-100053.zip](https://github.com/cilium/cilium/files/5303926/cilium-sysdump-20200930-100053.zip)

```bash
$ k8s-cilium-exec.sh cilium ip list       
IP                 IDENTITY   SOURCE
0.0.0.0/0          2          
10.233.64.45/32    104        k8s
10.233.64.49/32    104        k8s
10.233.64.70/32    4          
10.233.64.145/32   13684      k8s
10.233.64.163/32   21307      k8s
10.233.64.185/32   17504      k8s
10.233.64.221/32   104        k8s
10.233.64.244/32   4          
10.233.64.250/32   1          
10.233.65.55/32    1          
10.233.65.81/32    50316      k8s
10.233.65.85/32    11671      k8s
10.233.65.185/32   4          
10.233.65.226/32   4          
10.233.65.246/32   104        k8s
10.233.66.83/32    4          
10.233.66.172/32   4          
10.233.66.173/32   12910      k8s
10.233.66.187/32   41176      k8s
10.233.66.192/32   36116      k8s
10.233.66.220/32   40924      k8s
10.233.66.244/32   12842      k8s
10.233.66.245/32   3962       k8s
10.233.66.248/32   1          
169.254.25.10/32   16777218   
192.168.25.72/32   1          
IP                 IDENTITY   SOURCE
0.0.0.0/0          2          
10.233.64.45/32    104        k8s
10.233.64.49/32    104        k8s
10.233.64.70/32    4          
10.233.64.145/32   13684      k8s
10.233.64.163/32   21307      k8s
10.233.64.185/32   17504      k8s
10.233.64.221/32   104        k8s
10.233.64.244/32   4          
10.233.64.250/32   1          
10.233.65.55/32    1          
10.233.65.81/32    50316      k8s
10.233.65.85/32    11671      k8s
10.233.65.185/32   4          
10.233.65.226/32   4          
10.233.65.246/32   104        k8s
10.233.66.83/32    4          
10.233.66.172/32   4          
10.233.66.173/32   12910      k8s
10.233.66.187/32   41176      k8s
10.233.66.192/32   36116      k8s
10.233.66.220/32   40924      k8s
10.233.66.244/32   12842      k8s
10.233.66.245/32   3962       k8s
10.233.66.248/32   1          
169.254.25.10/32   16777218   
IP                 IDENTITY   SOURCE
0.0.0.0/0          2          
10.233.64.45/32    104        k8s
10.233.64.49/32    104        k8s
10.233.64.70/32    4          
10.233.64.145/32   13684      k8s
10.233.64.163/32   21307      k8s
10.233.64.185/32   17504      k8s
10.233.64.221/32   104        k8s
10.233.64.244/32   4          
10.233.64.250/32   1          
10.233.65.55/32    1          
10.233.65.81/32    50316      k8s
10.233.65.85/32    11671      k8s
10.233.65.185/32   4          
10.233.65.226/32   4          
10.233.65.246/32   104        k8s
10.233.66.83/32    4          
10.233.66.172/32   4          
10.233.66.173/32   12910      k8s
10.233.66.187/32   41176      k8s
10.233.66.192/32   36116      k8s
10.233.66.220/32   40924      k8s
10.233.66.244/32   12842      k8s
10.233.66.245/32   3962       k8s
10.233.66.248/32   1          
169.254.25.10/32   16777218   
192.168.25.70/32   1          
192.168.25.71/32   1          
```

**Comment by gandalfmagic** [comment]
Hello,
I found the problem, an it's not Cilium fault.

It seems that in Debian buster, with the latest backported versions of the linux kernels we need to upgrade a lot of packages to make things work.

To avoid problems systemd services (and ansible):

- systemd
- libnss-systemd
- libpam-systemd
- udev
- systemd-timesyncd
- sysv-rc
- initscripts
- insserv
- startpar

To avoid problems with BPF and cilium:

- iptables
- libiptc0
- libseccomp2
- nftables
- iproute2

After the upgrade, all the services works as intented.

**Comment by gandalfmagic** [comment]
**UPDATE**: my bad, the working solution was still on Kubernetes v1.17.13 with Cilium 1.8.4
Kubernetes 1.18.10 with Cilium 1.8.4 is still not working.

I've tested my cluster with cilium installed from kubespray with ansible and from the officil helm charts. No luck.

Thank you

**Comment by charles-woshicai** [comment]
Hello, I have got the same issue too, is there any update?

**Comment by joestringer** [comment]
From @charles-woshicai on Slack:

K8S version: v1.18.9
distro: Debian GNU/Linux 10 (buster)
Linux kernel version: 5.4.0-4-cloud-amd64
cilium version: v1.8 (it's installed by gardener, I don't know how to get this version)

**Comment by joestringer** [comment]
Thanks @gandalfmagic @charles-woshicai for the additional reports. These seem to indicate an issue specifically with k8s 1.18, though I can't think of what it could be just yet. We regularly test Cilium against k8s 1.18 and the CI is not picking up on any issue like this. 

I haven't had a chance to dig further into this just yet but I will be sure to update the thread if I find anything out.

**Comment by joestringer** [comment]
Something else I just noticed which may be quite relevant is that I see in the description `enable_nodelocaldns: true`. I'm not exactly sure what node-local-DNS will do in this scenario but it's plausible that this is somehow related to the failure.

**Comment by gandalfmagic** [comment]
> Something else I just noticed which may be quite relevant is that I see in the description `enable_nodelocaldns: true`. I'm not exactly sure what node-local-DNS will do in this scenario but it's plausible that this is somehow related to the failure.

I found that maybe is not a DNS issue only. Also the `toServices` rules have problems.

I use service with customized endpoints to regulate the traffic to external services, for example:

```yaml
kind: Service
apiVersion: v1
metadata:
  name: database
  namespace: myapp
spec:
  type: ClusterIP
  ports:
    - port: 3306
      targetPort: 3300
---
kind: Endpoints
apiVersion: v1
metadata:
  name: database
  namespace: myapp
subsets:
  - addresses:
      - ip: 10.0.0.1
    ports:
      - port: 3300
```

Then I use a `toServices` rule to permit the traffic from the application to the service:

```yaml
---
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: egress-myapp
  namespace: myapp
spec:
  endpointSelector:
    matchLabels:
      "k8s:app": myapp
  egress:
    - toServices:
        - k8sService:
            serviceName: database
            namespace: myapp
```

In Kubernetes 1.18, this rule fails the same way of the DNS resolution.

**Comment by charles-woshicai** [comment]
> > Something else I just noticed which may be quite relevant is that I see in the description `enable_nodelocaldns: true`. I'm not exactly sure what node-local-DNS will do in this scenario but it's plausible that this is somehow related to the failure.
> 
> I found that maybe is not a DNS issue only. Also the `toServices` rules have problems.
> 
> I use service with customized endpoints to regulate the traffic to external services, for example:
> 
> ```yaml
> kind: Service
> apiVersion: v1
> metadata:
>   name: database
>   namespace: myapp
> spec:
>   type: ClusterIP
>   ports:
>     - port: 3306
>       targetPort: 3300
> ---
> kind: Endpoints
> apiVersion: v1
> metadata:
>   name: database
>   namespace: myapp
> subsets:
>   - addresses:
>       - ip: 10.0.0.1
>     ports:
>       - port: 3300
> ```
> 
> Then I use a `toServices` rule to permit the traffic from the application to the service:
> 
> ```yaml
> ---
> apiVersion: cilium.io/v2
> kind: CiliumNetworkPolicy
> metadata:
>   name: egress-myapp
>   namespace: myapp
> spec:
>   endpointSelector:
>     matchLabels:
>       "k8s:app": myapp
>   egress:
>     - toServices:
>         - k8sService:
>             serviceName: database
>             namespace: myapp
> ```
> 
> In Kubernetes 1.18, this rule fails the same way of the DNS resolution.

Well, in my view it is the issue of `eBPF`, aftr applying `ciliumNetworkPolicy`, the DNS package is denied by `eBPF`. I couldn't find the log of `eBPF` so it's just my guess.  
Here is the command output in the pod created via `kubectl debug` which in the same linux network namespace.
```shell
bash-5.0# cat /etc/resolv.conf 
nameserver 100.64.0.10
search netpolicy.svc.cluster.local svc.cluster.local cluster.local eu-central-1.compute.internal
options ndots:5
bash-5.0# dig google.com @100.64.0.10

; <<>> DiG 9.14.12 <<>> google.com @100.64.0.10
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 3485
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: 02b9582b319cab21 (echoed)
;; QUESTION SECTION:
;google.com.			IN	A

;; ANSWER SECTION:
google.com.		30	IN	A	172.217.16.142

;; Query time: 1 msec
;; SERVER: 100.64.0.10#53(100.64.0.10)
;; WHEN: Fri Oct 23 02:27:54 UTC 2020
;; MSG SIZE  rcvd: 77

bash-5.0# dig google.com @8.8.8.8

; <<>> DiG 9.14.12 <<>> google.com @8.8.8.8
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 19481
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;google.com.			IN	A

;; ANSWER SECTION:
google.com.		299	IN	A	172.217.22.110

;; Query time: 10 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Fri Oct 23 02:28:49 UTC 2020
;; MSG SIZE  rcvd: 55

bash-5.0# # apply ciliumNetworkPolicy
bash-5.0# # could reach dns server both internal kube-dns and google dns
bash-5.0# dig google.com @100.64.0.10

; <<>> DiG 9.14.12 <<>> google.com @100.64.0.10
;; global options: +cmd
;; connection timed out; no servers could be reached
bash-5.0# dig google.com @8.8.8.8

; <<>> DiG 9.14.12 <<>> google.com @8.8.8.8
;; global options: +cmd
;; connection timed out; no servers could be reached
bash-5.0# 
```

**Comment by charles-woshicai** [comment]
@gandalfmagic I have some progress about this issue, thanks for the help from @joestringer .

Firstly run `cilium monitor -t drop`  inside the cilium container which is in the same node with your pod, and run `dig google.com`, here is the monitor output:
```shell
root@ip-10-250-4-151:/home/cilium# cilium monitor -t drop
Listening for events on 2 CPUs with 64x4096 of shared memory
Press Ctrl-C to quit
level=info msg="Initializing dissection cache..." subsys=monitor
xx drop (Policy denied) flow 0xf0433d07 to endpoint 0, identity 29995->1351: 100.96.3.106:51013 -> 100.96.3.61:8053 udp
xx drop (Policy denied) flow 0x684a7489 to endpoint 0, identity 29995->1351: 100.96.3.106:34037 -> 100.96.3.194:8053 udp
xx drop (Policy denied) flow 0xb61a3322 to endpoint 0, identity 29995->1351: 100.96.3.106:32842 -> 100.96.3.61:8053 udp
```

And I find `100.96.3.61:8053` is the the `endpoint` of `kube-dns` ep, which is strange.
```yaml
apiVersion: v1
kind: Endpoints
metadata:
  creationTimestamp: "2020-10-16T05:24:18Z"
  labels:
    k8s-app: kube-dns
    kubernetes.io/cluster-service: "true"
    kubernetes.io/name: CoreDNS
    shoot.gardener.cloud/no-cleanup: "true"
subsets:
- addresses:
  - ip: 100.96.3.194
    nodeName: ip-10-250-4-151.eu-central-1.compute.internal
    targetRef:
      kind: Pod
      name: coredns-7b8845bb85-t4hc5
      namespace: kube-system
      resourceVersion: "2378824"
      uid: 2f901ae8-475d-4f01-ae76-5cd38ced4110
  - ip: 100.96.3.61
    nodeName: ip-10-250-4-151.eu-central-1.compute.internal
    targetRef:
      kind: Pod
      name: coredns-7b8845bb85-fc9wt
      namespace: kube-system
      resourceVersion: "2876942"
      uid: 1248a761-05bc-4b4e-8d32-7585a686c204
  ports:
  - name: metrics
    port: 9153
    protocol: TCP
  - name: dns
    port: 8053
    protocol: UDP
  - name: dns-tcp
    port: 8053
    protocol: TCP
```
I try to edit the egress rule as below and run `dig google.com` in the pod, it works this time:
```yaml
spec:
  egress:
      toPorts:
        - ports:
            - port: '8053'
              protocol: ANY
          rules:
            dns:
              - matchPattern: '*'
```

There is still some problem to solve, hope this will help.

**Comment by joestringer** [comment]
Upon review of this issue, I believe that the original issue reported by @gandalfmagic is likely to be caused by the deployment of nodelocal DNS. Cilium v1.9.0 included native support for node-local DNS which can be deployed by following [the Node-LocalDNS cache guide](https://docs.cilium.io/en/v1.9/gettingstarted/local-redirect-policy/#node-local-dns-cache), so I would suggest trying that approach and reporting back if that does not work.

The latter half of this issue appears to have been more a confusion around which port must be used in DNS policy, which should be clarified by #14144.

**Comment by roodkcab** [comment]
for me, it's becase I use ipvlan instead of veth.

when using ipvlan, since toFQDNs needs dns proxy on cilium_host to update fqdn cache, all dns packages return to pod are redirect to dns proxy, which is not possible, cause bpf programm is located in pod and can't see cilium_host in root namespace of network.
 
theoretically, change back to veth should get the job done.

**Comment by pchaigno** [comment]
@joestringer Is there anything left to be done before we can close this issue? Your last comment seems to indicate the two root causes were addressed in the documentation and v1.9.0.

**Comment by joestringer** [comment]
Nothing left to close. I also posted a PR to clarify the point that @roodkcab pointed out above (#14893).
